include/devicetree/reset.h

```diff
@@ -0,0 +1,246 @@
+/**
+ * @file
+ * @brief Reset Controller Devicetree macro public API header file.
+ */
+
+/*
+ * Copyright (c) 2022, Andrei-Edward Popa
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DEVICETREE_RESET_H_
+#define ZEPHYR_INCLUDE_DEVICETREE_RESET_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup devicetree-reset-controller Devicetree Reset Controller API
+ * @ingroup devicetree
+ * @{
+ */
+
+/**
+ * @brief Get the node identifier for the controller phandle from a
+ *        "resets" phandle-array property at an index
+ *
+ * Example devicetree fragment:
+ *
+ *     reset1: reset-controller@... { ... };
+ *
+ *     reset2: reset-controller@... { ... };
+ *
+ *     n: node {
+ *             resets = <&reset1 10>, <&reset2 20>;
+ *     };
+ *
+ * Example usage:
+ *
+ *     RESET_DT_CTLR_BY_IDX(DT_NODELABEL(n), 0)) // DT_NODELABEL(reset1)
+ *     RESET_DT_CTLR_BY_IDX(DT_NODELABEL(n), 1)) // DT_NODELABEL(reset2)
+ *
+ * @param node_id node identifier
+ * @param idx logical index into "resets"
+ * @return the node identifier for the reset controller referenced at
+ *         index "idx"
+ * @see DT_PHANDLE_BY_IDX()
+ */
+#define RESET_DT_CTLR_BY_IDX(node_id, idx) \
```

Comment:
uhm, quite the opposite I think. To not pollute the `DT_` namespace we should use a different prefix.

Reply:
@carlocaione What about include/devicetree/pinctrl.h? All pinctrl DT apis begin with DT_.

Reply:
I'm looking at `include/drivers/pinctrl.h` because this is a driver.

Reply:
Ooook I see the problem now. These macros should be moved to `include/drivers/reset.h`

Reply:
There are 2 sides: the DT API part (include/devicetree/*) and the driver (include/drivers/*). AFAIK, the first needs needs `DT_`, the driver part `DRIVERNAME_`.

Reply:
right, but any good reason to not have everything in `include/drivers` like it was done for example for the `mbox` driver class?

Reply:
Because we want to create tests with this macros in tests/lib/devicetree/api maybe?

Reply:
you can still do that. Probably @mbolivar-nordic should have the final word here :)

Reply:
The current structure is OK if we want to have a DT API portion of this, but I am not sure if we do or not yet.

I have a few questions:

- same question as @gmarull about how this will work on platforms like STM32 that have a combined RCC (reset and clock control) peripheral; cc @erwango 
- is there a precedent for this sort of binding anywhere else? If we are working with a common paradigm inherited from Linux or something, I'll have fewer concerns, but if we are going off and inventing something new on our own, I'd want to know more about how this will generalize before extending the (stable) DT API

Reply:
> The current structure is OK if we want to have a DT API portion of this, but I am not sure if we do or not yet.
> 
> I have a few questions:
> 
> * same question as @gmarull about how this will work on platforms like STM32 that have a combined RCC (reset and clock control) peripheral; cc @erwango
> * is there a precedent for this sort of binding anywhere else? If we are working with a common paradigm inherited from Linux or something, I'll have fewer concerns, but if we are going off and inventing something new on our own, I'd want to know more about how this will generalize before extending the (stable) DT API

Okay, I will try to respond to that:

1)    I understand that some platforms a combined reset and clock control. I'm not really sure about how this works right now, the reset control part is used in some way with the clock API or only the clock control part is used? We can try to create 2 separate nodes in dts, one for clock control and the other one for reset control, both have the same reg address and use them separately. I'm not an expert, so I would gladly receive any suggestions for devices with RCC.
2)    Those bindings are more or less inherited from Linux. Linux uses a reset line 'unsigned long id' for information about resetting a device and this information can be passed directly into 'resets' binding for different devices. So I think that we didn't invent something new for this.

About DT API reset portion, I will modify it in the way that will be established.

Reply:
> Those bindings are more or less inherited from Linux. Linux uses a reset line 'unsigned long id' for information about resetting a device and this information can be passed directly into 'resets' binding for different devices. So I think that we didn't invent something new for this.

Can you give me a few pointers to how this works in the Linux source code?

Reply:
Yes, sure.
https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/reset/xlnx,zynqmp-reset.txt
https://elixir.bootlin.com/linux/latest/source/drivers/reset/reset-zynqmp.c
https://elixir.bootlin.com/linux/latest/source/arch/arm64/boot/dts/xilinx/zynqmp.dtsi#L816

Reply:
Thank you @andrei-edward-popa; this is very helpful.

I also see the following:

- https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/reset/reset.txt documents the general expectations for reset providers and consumers. This is consistent with the new API in include/devicetree/reset.h in this PR. I agree that if we are going to support bindings with `#reset-cells` and `resets` properties, we should have this API in devicetree.h, for consistency with other things like clocks, dmas, etc.
- the stm32 binding in https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/reset/st,stm32-rcc.txt is basically just a link to https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/clock/st,stm32-rcc.txt, which just documents that `#reset-cells` should be 1 for this IP, and documents [here](https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/clock/st,stm32-rcc.txt#L121) how to use this properly to set the 1 cell needed to identify a reset line on those SoCs from a consumer node
- https://elixir.bootlin.com/linux/latest/source/include/linux/reset-controller.h looks like it contains the driver API

Overall, the namespacing and organization of the DT macros look good to me; consider this message my +1 on the approach from the devicetree side.

The existence of a corresponding driver API in zephyr also sounds like it could be useful, but I haven't thought about it much. The design and implementation of the higher-level driver APIs I will leave to you all to figure out :).

Reply:
> Overall, the namespacing and organization of the DT macros look good to me; consider this message my +1 on the approach from the devicetree side.

Is it though? If we want to keep the DT macros in `include/devicetree/reset.h` then I think the namespace should be `DT_RESET_*` instead of `RESET_DT_*` to be consistent with what we do with the other subsystems.

Reply:
@carlocaione Check the commits, right now it is the version with DT_RESET.